markdown/util.py

```diff
@@ -44,7 +44,298 @@
+        self._set.discard(element)
+        try:
+            self._list.remove(element)
+        except ValueError:
+            pass
+
+    def extend(self, elements: list[str], /) -> None:
+        warnings.warn(
+            "Using block level elements as a list is deprecated, use it as a set instead.",
+            DeprecationWarning,
+        )
+        # In-place extension should update both list and set.
+        self._list.extend(elements)
+        self._set.update(elements)
+
+    def index(self, value, start: int = 0, stop: int = sys.maxsize, /):
+        warnings.warn(
+            "Using block level elements as a list is deprecated, use it as a set instead.",
+            DeprecationWarning,
+        )
+        return self._list.index(value, start, stop)
+
+    def insert(self, index: int, element: str, /) -> None:
+        warnings.warn(
+            "Using block level elements as a list is deprecated, use it as a set instead.",
+            DeprecationWarning,
+        )
+        # In-place insertion should update both list and set.
+        self._list.insert(index, element)
+        self._set.add(element)
+
+    def intersection(self, *others: set[str]) -> set[str]:
+        # User expects a set back.
+        return self._set.intersection(*others)
+
+    def intersection_update(self, *others: set[str]) -> None:
+        # In-place intersection should update both list and set.
+        self._set.intersection_update(*others)
+        # Elements were only removed.
+        self._list[:] = [element for element in self._list if element in self._set]
+
+    def isdisjoint(self, other: set[str], /) -> bool:
+        return self._set.isdisjoint(other)
+
+    def issubset(self, other: set[str], /) -> bool:
+        return self._set.issubset(other)
+
+    def issuperset(self, other: set[str], /) -> bool:
+        return self._set.issuperset(other)
+
+    def pop(self, index: int | None = None, /) -> str:
+        # In-place pop should update both list and set.
+        if index is None:
+            index = -1
+        else:
+            warnings.warn(
+                "Using block level elements as a list is deprecated, use it as a set instead.",
+                DeprecationWarning,
+            )
+        element = self._list.pop(index)
+        # Only remove from set if absent from list.
+        if element not in self._list:
+            self._set.remove(element)
+        return element
+
+    def remove(self, element: str, /) -> None:
+        # In-place removal should update both list and set.
+        # We give precedence to set behavior, so we remove all occurrences from the list.
+        while True:
+            try:
+                self._list.remove(element)
+            except ValueError:
+                break
+        # We raise `ValueError` for backwards compatibility.
+        try:
+            self._set.remove(element)
+        except KeyError:
+            raise ValueError(f"{element!r} not in list") from None
```

Comment:
Not sure about this. It's difficult to deprecate a raised exception in favor of another one. Maybe we should immediately start raising `KeyError`.

Reply:
I'm inclined to suggest that we not bother ourselves with the errors here. We can just call `_list.remove` and `_set.remove` and allow any errors to bubble up to the calling code. Is there any reason why we shouldn't do that?

Reply:
Agree on keeping things simple. I'll try a different approach (we still want to remove *all* matching items from the list, not just one) :thinking: